src/System.Text.Primitives/System/Text/FormattingData.cs

```diff
@@ -12,19 +12,309 @@ public struct FormattingData
+        public enum ParsingStates
+        {
+            NO_PARSING = 0, // don't construct it
+            PARSING = 1,    // construct it with upfront cost
+            LAZY_PARSING = 3 // construct it but allocate lazy. 3 is chosen so the bitwise check below work
+        }
 
-        public FormattingData(byte[][] digitsAndSymbols, Encoding encoding)
+        public FormattingData(byte[][] digitsAndSymbols, Encoding encoding, byte parsing = (byte)ParsingStates.NO_PARSING)
         {
             _digitsAndSymbols = digitsAndSymbols;
             _encoding = encoding;
+            if (((byte)parsing & (byte)ParsingStates.PARSING) == 1) // both PARSING and LAZY_PARSING pass this check
+            {
+                _parsingTrieList = new List<TrieNode>();
+                _childTrieNodes = new List<TrieNode>();
+                ConstructParsingTrie(parsing == (byte)ParsingStates.LAZY_PARSING);
+            }
+            else
+            {
+                _parsingTrieList = null;
+                _childTrieNodes = null;
+            }
+        }
+        /// <summary>
+        /// Construct a formatting data by specifying the bytes for each value
+        /// </summary>
+        /// <param name="digit0">The bytes which map to the digit zero (0)</param>
+        /// <param name="digit1">The bytes which map to the digit one (1)</param>
+        /// <param name="digit2">The bytes which map to the digit two (2)</param>
+        /// <param name="digit3">The bytes which map to the digit three (3)</param>
+        /// <param name="digit4">The bytes which map to the digit four (4)</param>
+        /// <param name="digit5">The bytes which map to the digit five (5)</param>
+        /// <param name="digit6">The bytes which map to the digit six (6)</param>
+        /// <param name="digit7">The bytes which map to the digit seven (7)</param>
+        /// <param name="digit8">The bytes which map to the digit eight (8)</param>
+        /// <param name="digit9">The bytes which map to the digit nine (9)</param>
+        /// <param name="decimalSeparator">The bytes which map to the decimal separtor.</param>
+        /// <param name="groupSeparator">The bytes which map to the group separator</param>
+        /// <param name="infinity">The bytes which map to the representation of infinity</param>
+        /// <param name="minusSign">The bytes which map to the minus sign</param>
+        /// <param name="plusSign">The bytes which map to the plus sign</param>
+        /// <param name="nan">The bytes which map to the representation of NaN (not a number)</param>
+        /// <param name="exponent">The bytes which map to the exponent marker for floating-point types</param>
+        /// <param name="encoding">The specified encoding</param>
+        /// <param name="parsing"></param>
+        public FormattingData(byte[] digit0, byte[] digit1, byte[] digit2, byte[] digit3, byte[] digit4, byte[] digit5, byte[] digit6, byte[] digit7,
```

Comment:
Why are you adding this constructor? Take a look into System.Formatting.Globalization for samples (tests) showing how to construct non-invariant cultures.